### ISSUE TITLE
generic snapshot - documentation

### ISSUE THREAD

**Issue by brodybits** [15r10nk/inline-snapshot#49]
Hey thanks for giving us such an awesome tool - I use it almost every day in my own unit testing work!

I would be interested in support for generic snapshots. Here is some code that illustrates what does NOT work, as already documented, and an undocumented workaround I was able to figure out:

```python
from inline_snapshot import snapshot

from types import SimpleNamespace


def bold(text):
    return f"<b>{text}</b>"


def test_bold():
    assert bold('word') == snapshot("<b>word</b>")

    assert bold('__generic__text__') == _generic_bold_snapshot

    replacement_text = 'replacement text'

    # DOES NOT WORK - BUT WILL WORK WITH --inline-snapshot-disable option flag:
    # assert bold(replacement_text) == _generic_bold_snapshot.replace('__generic__text__', replacement_text)

    # USING UNDOCUMENTED WORKAROUND - WILL FAIL WITH --inline-snapshot-disable option flag:
    assert bold(replacement_text) == _generic_bold_snapshot.get_result(SimpleNamespace(create=False, fix=False)).replace('__generic__text__', replacement_text)


_generic_bold_snapshot = snapshot("<b>__generic__text__</b>")
```

(I would more likely be using a generic snapshot with list of strings in my work.)

I think the undocumented workaround is very clearly hackish and potentially subject to easy breakage.

A nicer, documented API for this would be really awesome.

I may be able to work on a proposal someday, unfortunately cannot make any promises at this point.

Thanks and thanks again!

---

P.S. My results would change if I would test with the --inline-snapshot-disable flag, and I have updated a couple test comments with this info.

**Comment by 15r10nk** [comment]
Great that you like it.

It is the intention that snapshots cannot be read (because it should not be necessary).
You can use the value which you compare with the snapshot instead.

The following code should work for you.

``` python
from inline_snapshot import snapshot


def bold(text):
    return f"<b>{text}</b>"


def test_bold():
    assert bold("word") == snapshot("<b>word</b>")

    generic_bold = bold("__generic__text__")

    assert generic_bold == snapshot("<b>__generic__text__</b>")

    replacement_text = "replacement text"

    assert bold(replacement_text) == generic_bold.replace(
        "__generic__text__", replacement_text
    )
```

I do not expose the value which is stored inside the snapshot because it difficult to know which value the user expects.
* the current value which is written in the source code
* the value after a potential fix

I think the code is easier to read if you use the values which you already have and where you can easily reason about their values.

Please let me know if the code above works for you.

**Comment by brodybits** [comment]
Makes sense, thanks. I would personally favor keeping the generic snapshot at the end, as it could be over 100 lines with a list of tuples and/or dicts with the generic patterns to be replaced in my case.

Here is an example that shows how that kind of generic comparison could be done with a function that returns a list of strings:

```python
from inline_snapshot import snapshot


# ================================================

# SOURCE TO BE TESTED


def work(text):
    return [f"{text}.html", f"<b>{text}</b>"]


# ================================================

# SINGLE NON-GENERIC TEST CASE


def test_work_with_single_word_inline():
    assert work("word") == snapshot(["word.html", "<b>word</b>"])


# ================================================

# TEST WITH GENERIC REPLACEMENT IN ITS OWN MODULE


def test_work_with_replacement_text():
    replacement_text = "replacement text"

    assert work(replacement_text) == get_expected_work_result(replacement_text)


# ================================================

# FOR A GENERIC TEST MODULE


def get_expected_work_result(text):
    return [x.replace("__generic__text__", text) for x in _generic_work_result]


# ---


def _get_generic_work_result():
    return work(__GENERIC__TEXT__)


# ---

__GENERIC__TEXT__ = "__generic__text__"

# ---


def test_generic_work_result():
    generic_work = work(__GENERIC__TEXT__)

    assert generic_work == _generic_work_snapshot


# ---

_generic_work_result = _get_generic_work_result()

_generic_work_snapshot = snapshot(
    ["__generic__text__.html", "<b>__generic__text__</b>"]
)
```

In the real world I would likely want to alphabetize the test cases & test file names to be sure to test & update the generic test cases before attempting any of the tests using the text replacement. I may also use pytest `-x` or `--maxfail=...` to avoid the extra noise in case of any generic test failures.

I would love to see this documented better, may try to contribute something (someday).

**Comment by 15r10nk** [comment]
You could use something like [pytest-depends](https://pypi.org/project/pytest-depends/) (I have no personal experience using it) or you could use a fixture like this:

``` python
from inline_snapshot import snapshot


# ================================================

# SOURCE TO BE TESTED


def work(text):
    return [f"{text}.html", f"<b>{text}</b>"]


# ================================================

# SINGLE NON-GENERIC TEST CASE


def test_work_with_single_word_inline():
    assert work("word") == snapshot(["word.html", "<b>word</b>"])


# ================================================

# TEST WITH GENERIC REPLACEMENT IN ITS OWN MODULE


def test_work_with_replacement_text1(get_expected_work_result):
    replacement_text = "replacement text 1"

    assert work(replacement_text) == get_expected_work_result(replacement_text)

def test_work_with_replacement_text2(get_expected_work_result):
    replacement_text = "replacement text 2"

    assert work(replacement_text) == get_expected_work_result(replacement_text)

# ================================================

# THE PYTEST FIXTURE

import pytest

__GENERIC__TEXT__ = "__generic__text__"


@pytest.fixture(scope="session")
def get_expected_work_result():
    print("fixture is only run one time because of the session scope")

    _generic_work_result = work(__GENERIC__TEXT__)

    assert _generic_work_result == _generic_work_snapshot

    def get_result(text):
        return [x.replace("__generic__text__", text) for x in _generic_work_result]

    return get_result


_generic_work_snapshot = snapshot(
    ["__generic__text__.html", "<b>__generic__text__</b>"]
)
```

> I would love to see this documented better, may try to contribute something (someday).

I hear you. inline-snapshot is a framework which allows you to create your own way how you want to test your software (which you do and I like this).

I will add an examples section to the documentation with some common pattern how inline-snapshot can be used. I hope this will provide some guidance for using inline-snapshot in more advanced ways.

I also want to thank you for your support. This motivates me to continue my work :smiley: .

**Comment by brodybits** [comment]
Thanks and thanks for the info in the other issues I raised.

**Comment by brodybits** [comment]
Apologies I hit the wrong button (still getting used to a new laptop I just started using) ... `@pytest.fixture` does seem to make it much more readable than what I made, and using `pytest-depends` could be an nice improvement as well.

At this point I may put the generic snapshot on hold in favor of another, though likely dumber approach.

I am definitely looking forward to the updated documentation, whenever (and if ever) you get the chance to update it. And my contribution was meant only as a small token of appreciation, does not come close to justice for how much this has helped me with my work. I am hoping to get something more from the client but cannot promise anything for anytime soon (if ever). Sure hope this project will get some more updates, even if they are very small. Thanks again in any case.
